[WIP] Redesigned Validations

siemensikkema opened the pull request:
This redesign offers better options for localization of validation errors, a simpler syntax with nested validations, and experimental function builder support.

All validation functionality is defined on `Sequence` of `Validation` removing the need for a mutable `Validations` container.

# Nested validations
```swift
let validations = [
    Validation(key: "name", as: String.self, validator: .count(5...10)),
    Validation(key: "missing", validator: .ascii),

    // add an embedded validatable model
    Validation(key: "embedded1", validatable: Test.Embedded.self),

    // or alternatively, define the embedded validations in place
    Validation(key: "embedded2", validations: [
        Validation(key: "string1", as: String.self, validator: .count(3...)),
        Validation(key: "email", validator: .email)
    ])
]
try validations.validate(data, using: JSONDecoder())
```

# Localization (example)

Improved options for localization is achieved by exposing the validation failures as separate types so they can be extended. An example with some `Localizable` protocol:

```swift
extension CountValidatorFailure: Localizable {
    public var localization: Localization {
        switch self {
        case let .greaterThan(max):
            return .init(key: "validations.greaterThan", substitutions: ["%d": "\(max)"])
        case let .lessThan(min):
            return .init(key: "validations.lessThan", substitutions: ["%d": "\(min)"])
        }
    }
}
```

Producing a localized message could then look like this:

```swift 
do {
    try validations.validate(data, using: JSONDecoder())
} catch(let error as ValidationsError) {
    print(error.describeFailures {
        ($0 as? Localizable).map(localizer.localize)
    })
}
```

# Function Builders
With function builders the above validation could look like:

```swift
try Validations {
    Validation(key: "name", as: String.self, validator: .count(5...10))
    Validation(key: "missing", validator: .ascii)

    // add an embedded validatable model
    Validation(key: "embedded1", validatable: Test.Embedded.self)

    // or alternatively, define the embedded validations in place
    Validation(key: "embedded2") {
        Validation(key: "string1", as: String.self, validator: .count(3...))
        Validation(key: "email", validator: .email)
    }
}
```

Misc:
- No “readable” on validators or “description” on validation failure. Up to the developer how to represent them.
- Suggestion: make Validations a separate, Vapor agnostic, package, with a few simple extensions  (see `Validations+Vapor.swift`) in the Vapor package to make it compatible with `Request` and such.
- Suggestion: use BasicCodingKey (with the added initializer) and DecoderUnwrapper from CodableKit.
- The Validatable protocol is not required for using Validations. All functionality is based around sequence of Validation. This allows for flexibility in defining the validations for the current context. 
- There are some function builder helpers but they are not required for using Validations. One could add @ValidationBuilder to the validations() function when implementing Validatable or using the global function `Validations` (capitalization intended) but that only saves you from explicitly creating an array. One downside is that you can’t use the function builders at the moment for create a single validation, only >= 2.
- Validations can now be defined in a nested way and as such do not need to decode a subcontainer several times if there are multiple validatable properties in a contained object.
- The CompoundValidatorFailure is used for validators that work on other validators (= “higher order validators”) like `and`, `or`, `not`. I’m not sure this should stay as is because for proper description of the failed validations we might need to know what validator caused the failure.
- I made the Failures public but the Validators internal.
- For the NotValidator I decided to include the failure type of the succeeding validator since I removed `validatorReadable`. With the right extensions to ValidatorFailure this should make it possible to create a (localized) description.
- I improved the EmailValidator with a regex that is closer to a complete email validation. Perhaps an email parser is required to be sure the full spec is adhered to.

### Checklist
- [ ] Circle CI is passing (code compiles and passes tests).
- [ ] There are no breaking changes to public API.
- [ ] New test cases have been added where appropriate.
- [ ] All new code has been commented with doc blocks `///`.

siemensikkema commented on Sources/Vapor/Validation/Validator.swift:
```diff
@@ -1,38 +1,16 @@
-    /// - throws: `ValidationError` if the data is not valid, or another error if something fails.
-    func validate(_ data: ValidationData) -> ValidatorFailure?
+    associatedtype Data: Decodable
+    associatedtype Failure: ValidatorFailure
```
`ValidatorFailure` currently is an empty protocol without inheriting from `Error`. It's really only there to avoid relying on `Any`. I'm not sure we should require the Failures to conform to Error. That might be confusing because we never `throw` them.

siemensikkema commented:
@tanner0101 thanks a lot for your detailed reply!

> What about if we publicize the validators as well, but under the `Validator` namespace. Then we can nest the failures inside their respective validators:
> 
> * `Validator.Nil.Failure`
> * `Validator.Count.Failure`
> * etc

Great suggestion, I have made this change.

> While I do like the idea of relying more on stdlib types (`Sequence where Element == Validation`), I think the new API is more verbose.  ...

I'd like to try and see if we can combine the expressivity with the simplicity of "Sequence of Validation". We can replace `validator` with `is` to get closer to the original syntax.

```swift
return [
    Validation(key: "name", as: String.self, is: .count(5...) && .alphanumeric),
    Validation(key: "age", as: Int.self, is: .range(18...))
]
```

We can get shorter still by relying on type inference:

```swift
[
    .init(key: "name", as: String.self, is: .count(5...) && .alphanumeric),
    .init(key: "age", as: Int.self, is: .range(18...))
]
```

Maybe even remove "key" since that is clear from the context

```swift
[
    .init("name", as: String.self, is: .count(5...) && .alphanumeric),
    .init("age", as: Int.self, is: .range(18...))
]
```

We _could_ go even further and introduce operators:

```swift
[
    "name" <~ .count(5...) && .alphanumeric, // type can be inferred here
    "age" <~ Validator<Int>.range(18...) // have to specify type
]
```

With `<~` for required fields and `<~?` for optional ones.

> Misc stuff:
> I agree that the developer should be able to decide how they are represented, but I also think Vapor should provide a reasonable default here. The validation error messages we have currently on `master` are still very useful for developers looking at internal errors and console logs.

I'll bring the descriptions back while keeping the customization.

> > Suggestion: make Validations a separate, Vapor agnostic, package, with a few simple extensions (see Validations+Vapor.swift) in the Vapor package to make it compatible with Request and such.
> 
> I agree. We can and should do this as a separate, future PR

👍 

> > Suggestion: use BasicCodingKey (with the added initializer) and DecoderUnwrapper from CodableKit.
> 
> While I'm all for code re-use, I think these few small hacks are fine being reimplemented. Unless we end up needing more from CodableKit in the future.

👍 

> I think function builders here is interesting, but I'm not yet convinced it's better than the `Validations` API on master. Maybe it would be better as something like this:
> 
> ```swift
> Validations {
>     Validation(on: "name", as: String.self) {
>         Count(in: 5...) && Alphanumeric()
>     }
>     Validation(on: "age", as: Int.self) {
>         Range(in: 18...)
>     }
>     Validation(nested: "pet") {
>         Validation(on: "name", as: String.self) { 
>             Count(in: 5...) && CharacterSet(in: .alphanumerics + .whitespaces)
>         }
>         Validation(on: "pet", as: Int.self) { Range(in: 3...) }
>     }
> }
> ```

I agree that the function builder stuff doesn't add much beyond removing some comma's. If we end up choosing the operators it could look pretty nice though:

```swift
Validations {
    "name" <~ .count(5...) && .alphanumeric
    "age" <~ Validator<Int>.range(18...)
    "pet" <~ Validations {
        "name" <~ .count(5...)
    }
}
```

siemensikkema commented:
I currently have disabled the `Not` validator because it doesn't work well without `validatorReadable`. In other words, since the proposed implementation depends on the existence of Failure to describe validation failures, how do we describe the absence of an expected failure?
We'd need to either make Validators self-describing again (including localization) or consider an alternative approach like:

`.not(.ascii, orFailWith: MyCustomLocalizableValidationFailure())`

Although its syntax on `master` is really nice, in practice I haven't found that much use for the `NotValidator` so I don't think it is a huge loss.

tanner0101 commented on Sources/Vapor/Validation/Validators/In.swift:
```diff
@@ -0,0 +1,43 @@
+extension Validator {
+    /// Validates whether an item is contained in the supplied array.
+    public struct In<T: Decodable & Equatable>: ValidatorType {
+        public struct Failure: ValidatorFailure {
```
Hmm I see a problem with this idea now. Even though `Validator.In.Failure` is not actually generic on `T`, you can only reference it if you choose a specific `T`. So if you wanted to cast some `ValidatorFailure` as this, you'd need to do:
```swift
failure as? Validator.In<Int>.Failure`
```
You'd need to do that for every possible type, too. We also have to make the Validator public which is not great since on `master` we can keep that all internal. 

What about if we nest the failure types under just `Validator`? So for this case, we'd have:
```swift
extension Validator { 
    public struct InFailure: ValidatorFailure { ... }
}
```

siemensikkema replied:
Casting like that is indeed impractical. However, that's not how I would approach something like localization. Instead you can just conform the Failure to a protocol like so in the case of `CustomStringConvertible`:
```swift
extension Validator.In.Failure: CustomStringConvertible {
  ...
}
```
then, at the point where you needed the failure's custom string implementation you would just cast it like
```swift
failure as? CustomStringConvertible
```
---
Having said all that I've been able to make all validators non-generic, relying only on the `Validator`'s `T`.
---
I'll try to make the validators internal again and only the failures public in a later commit.

tanner0101 commented on Sources/Vapor/Validation/Validators/And.swift:
```diff
@@ -0,0 +1,24 @@
+            self.rhs = rhs
+        }
+
+        public func validate(_ data: T) -> CompoundValidatorFailure? {
```
What's the reason for using `CompoundValidatorFailure` here? I was expecting `And` to have its own `ValidatorFailure` like the other validators do. 

Something like:
```swift
enum AndValidatorFailure: ValidatorFailure {
    case left(ValidatorFailure)
    case right(ValidatorFailure)
    case both(left: ValidatorFailure, right: ValidatorFailure)
}
```

siemensikkema replied:
It came from the need to "flatten" the validator failures for these "higher order" validators when creating the descriptions. But I think we should stick with one failure type for each validator as you suggest.

tanner0101 commented on Sources/Vapor/Validation/ValidationsError.swift:
```diff
@@ -0,0 +1,33 @@
+        self.failedValidations = failedValidations
+    }
+
+    public func describeFailures(_ describe: (ValidatorFailure) -> String?) -> [(path: String, failures: [String])] {
```
What about just making `failedValidations` public? Then we don't need this method.

tanner0101 commented on Sources/Vapor/Validation/Validation.swift:
```diff
@@ -0,0 +1,104 @@
+        self.init(key: BasicCodingKey(key), required: required, validator: validator)
+    }
+
+    public init(key: BasicCodingKey, required: Bool = true, validations: [Validation]) {
```
What's the benefit of supporting both `BasicCodingKey`, `CodingKey`, and `String` for each method? I think just `String` should be sufficient for all use cases.

siemensikkema replied:
I made it this way because in some cases you might have the CodingKey already available and you can eliminate typos. I'll remove it for now for simplicity's sake. You can still use CodingKeys and take their String values.

tanner0101 commented on Sources/Vapor/Validation/Validators/Not.swift:
```diff
@@ -0,0 +1,21 @@
+///// Inverts a `Validation`.
+//public prefix func !<T: Decodable> (validator: Validator<T>) -> Validator<T> {
```
If the errors did not include `is` in the description, then we would be able to implement not by just prepending `not` to the description. Then finally when `ValidationErrors` prints out the errors, it could prepend `is`.

siemensikkema replied:
The problem here is that Validators (in this PR) do not have a way to describe themselves. So if the NotValidator fails, that means there is no ValidatorFailure to base the description of. I see 3 solutions:
- bring back `validatorReadable` (haven't needed this outside inverted validators)
- introduce `.not(.someValidator, ofFailWith: MyCustomValidatorError())` syntax
- add a requirement to ValidatorFailure to describe its inverted state. I prefer this and will try this out to see how it feels working with it.

tanner0101 replied:
> add a requirement to ValidatorFailure to describe its inverted state. I prefer this and will try this out to see how it feels working with it

Would this be a static requirement? I'm not sure that would work then since the information about what failed is stored in the error created at runtime.

I think for this to work we need to have a concept of a "ValidationSuccess".

siemensikkema commented:
> ```swift
> [
>     .init("name", as: String.self, is: .count(5...) && .alphanumeric),
>     .init("age", as: Int.self, is: .range(18...))
> ]
> ```
> 
> This seems a little better. I still think having a distinct `Validations` type could come in handy in the future though. Say for example you want to add some "validation metadata":
> 
> ```swift
> static func validations(_ v: inout Validations) {
>     v.add("name", as: String.self, is: .count(5...))
>     v.add("age", as: Int.self, is: .range(18...))
>     v.acceptedContentTypes = [.json, .xml]
> }
> ```
> 
> I'm not sure how likely anything like that would be. But just using an array and extensions does seem a bit restricting.

I get where you're coming from but I don't think this example is something I would use validations for. Besides, I think Validations, once it is separate from Vapor, will have the `func validate(_ request: Request) throws` methods meaning the content type is not available. 
I think the restricted nature is actually a good thing but since Validation.ValidationType is not public we're be free to add other validation types later as long as they operate on something we can derive from `Decoder` or `KeyedDecodingContainer<BasicCodingKey>`.

tanner0101 commented:
> Besides, I think Validations, once it is separate from Vapor, will have the func validate(_ request: Request) throws methods meaning the content type is not available.
I think the restricted nature is actually a good thing but since Validation.ValidationType is not public we're be free to add other validation types later as long as they operate on something we can derive from Decoder or KeyedDecodingContainer<BasicCodingKey>.

If we had a distinct `Validations` struct we could offer something like `userInfo` on it that would make it extensible. Then when `ValidationKit` (or whatever we call it) is imported into Vapor, Vapor could add properties and methods to it to include things like `acceptedContentTypes`.

tanner0101 commented:
Expanding on my thoughts about fixing descriptions for the `Not` validator, maybe we can generalize `ValidationFailure` as `ValidationResult`?

```swift
protocol ValidationResult: CustomStringConvertible {
    var failed: Bool { get }
}
```

If `Validator` always returns a `ValidationResult`, then `Not` will have the information it needs to return a readable string about the result. To check if validations failed overall, we just do: `validations.filter { $0.failed }.isEmpty`. The information about succeeded validations could be useful for logging, too

siemensikkema commented:
> Expanding on my thoughts about fixing descriptions for the `Not` validator, maybe we can generalize `ValidationFailure` as `ValidationResult`?

Genius. It's so obvious now! I've been trying it out and it mostly falls into place. There are still some gotchas when implementing how to negate result descriptions. For example, what if `!.empty` fails? That would result in "not not empty" in my current implementation, which, although correct, is not exactly ideal 😆

tanner0101 commented:
> Genius. It's so obvious now! I've been trying it out and it mostly falls into place.

Sweet!

> For example, what if !.empty fails? That would result in "not not empty" in my current implementation, which, although correct, is not exactly ideal 😆

Haha. That is ... interesting. Hacky solution would be to find and replace `"not not"` => `""`. I wonder if we can do better though.

siemensikkema commented:
@tanner0101 Hi again, it's been a while ... 🙂 Let me show you what I've worked on since we last spoke here.

- made Validators internal again and scoped ValidatorResults under Validator
- remove non-string key initializers from Validation
- Implemented the concept of ValidatorResult with a boolean property `failed`. This allows us to log information about succeeded validations as well as failed ones. Note that the problem with the `Not` validator was solved by ...
- requiring types conforming to `ValidatorType` to be able to create an inverted version of themselves (`.inverted()`). This can either create a copy of itself with a flag flipped or a different type. For example, in case of the `And` validator the negated version is an `Or` with the contained validators negated according to De Morgan's law (and vice versa for `Or`). This helps greatly in creating messages that flow nicely in every situation because the ValidatorResult now also has the full responsibility for creating the descriptions (so no more "is" and "not" prepended). We also do not have to deal with double negation in hacky ways like (like removing "not not ").
- updated the syntax of the validation initializers to match what was discussed.
- added the ability to add an existing validation result. This adds support for cases where we cannot determine the validation result from a decoder alone. This includes for example a request's `contentType` and asynchronous validations. To determine whether an email is unique we could do something like this:
```swift
struct UniqueValidatorResult {
    let fieldName: String
    let description: String {
        "\(fieldName is \(failed ? "not " : "") unique"
    }
    let failed: Bool
}
let email: String = try req.content("email")
req.db.query(User.self).filter(\.email == email).first().flatMapThrowing { existingEmail in
    let validations = User.validations() + [Validation("email", result: UniqueValidatorResult(fieldName: "email", failed: true)]
    return try validations.validate(req)
}.map {
    ...
}
```
- removed function builder stuff

I have not (yet) reintroduced the Validations type because I think the new initializer for existing validator results covers the same functionality. In both implementations we are still restricted to associating a validator failure to a path. I'm not sure that fits the use case of a request's contentType very well so that kind of validation mighty fall outside the scope. Unless we change that restriction somehow.

Magiguigui commented:
Why it is not in https://github.com/vapor/validation ? Does some frameworks are now nested in vapor ?

There is some good improvements made here, good job ! As express above I think it could be a good idea to access request or only a dbconnection into validation process to open the possibilities while validating (unique email for example, in the current version (2.x.x) of Validation it is impossible to check)

siemensikkema commented:
@Magiguigui This branch has since been taken over by @tanner0101 and included in master IIRC. I believe it is in this repo only for efficiency in development and should end up back in vapor/validation some time during/after the beta period.
Your example is exactly one of those that we tried to solve with https://github.com/nodes-vapor/submissions for Vapor 3. My hope is that the new Validation package will make most of Submissions obsolete. I think email uniqueness (or any other db/async validation) can be achieved by first performing the lookup asynchronously and with the result build up the validations. That means you cannot use a Validatable type (since db access is not possible there) but you don't need to, since validations can be made anywhere.
Note that Tanner's version of the implementation is not based on "Sequence of Validation" but uses a Validations type which wraps an array of Validations.

Magiguigui commented:
Nice

Magiguigui commented:
It is a bad idea to make all validation run async, so like that we could build async validations next to "basic" validations and keep consistency... If we do so we will be able to pass a Request object to the validations method (Validatable) and run every kind of validations at the same place ?

siemensikkema commented:
@Magiguigui I think we should not involve the request. That can be up to the developer. I'm more for a simple API that can be used as you see fit. If that involves making all validations async because your use-case demands it then that should be possible but not required IMO.

Magiguigui commented:
Good point. What I did in finé is to create an AsynValidatorType protocol, new AsyncValidator struct and there is an example of unique async validator : 
```swift
public struct UniqueAsyncValidator<U: Model>: AsyncValidatorType {

	public var asyncValidatorReadable: String = "a unique field"

	private let keyPath: KeyPath<U, String>

	public init(at keyPath: KeyPath<U, String>) {
		self.keyPath = keyPath
	}

	public func validate(_ data: String, on request: Request) throws -> Future<Void> {
		return U
			.query(on: request)
			.filter(keyPath, "=", value)
			.count()
			.map { count in
				guard count == 0 else {
					throw BasicValidationError("not unique")
				}
			}
	}

}
```

Magiguigui commented:
I agree, I still think for user related use-cases we should be able to access Request object if we need it : 
With the current API architecture it's already possible by using a custom Validatable protocol. Why don't add it in the defaults implementation ? (it's only few changes : see below)
I can PR for that of course (This proposal is generic enough and not that much related to my use case I hope...)

```swift
public protocol Validatable {
    static func validations(_ validations: inout Validations, on request: Request)
}

extension Validatable {
    public static func validate(_ request: Request) throws {
        try self.validations(request).validate(request).assert()
    }
    
    public static func validate(json: String) throws {
        try self.validations().validate(json: json).assert()
    }
    
    public static func validate(_ decoder: Decoder) throws {
        try self.validations().validate(decoder).assert()
    }
    
    public static func validations(_ request: Request) -> Validations {
        var validations = Validations()
        self.validations(&validations, on: request)
        return validations
    }
}

```
